6.analyze_features/scripts/8a.mean_aggregated_PCA_calulate_mahalanobis_distance.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This notebooke calculates the mahalnobis distance between points on a pca.
+# I will document more about what mahalnobis distance is and how it is calculated in this notebook.
+
+# In[1]:
+
+
+import itertools
+import pathlib
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import mahalanobis
+
+# In[2]:
+
+
+# set the path to the data
+mean_aggregated_data_pca_path = pathlib.Path(
+    "../../data/6.analysis_results/mean_aggregated_pca.parquet"
+).resolve(strict=True)
+
+# read the data
+mean_aggregated_data_pca = pd.read_parquet(mean_aggregated_data_pca_path)
+print(mean_aggregated_data_pca.shape)
+mean_aggregated_data_pca.head()
+
+
+# # Mahalanobis Distance
+# For more in depth information on Mahalanobis distance, please refer to this [link](https://medium.com/@the_daft_introvert/mahalanobis-distance-5c11a757b099).
+# Mahalanobis distance is a measure of the distance between a point P and a distribution D.
+# It is a multi-dimensional generalization of the idea of measuring how many standard deviations away P is from the mean of D.
+# This distance is zero if P is at the mean of D, and grows as P moves away from the mean along each principal component axis.
+# The formula for Mahalanobis distance is given by:
+# ### $D^2 = (x - \mu)^T \Sigma^{-1} (x - \mu)$
+# where:
+# - $D$ is the Mahalanobis distance
+
+# ## Caluclating Mahalanobis Distance on PCA components 1 and 2
+# Here I calculate the Mahalanobis distance between the points on the first two principal components of the data for each of the three genotypes.
+
+# In[3]:
+
+
+# split the genotypes
+high_severity = mean_aggregated_data_pca[
+    mean_aggregated_data_pca["Metadata_genotype"] == "High-Severity"
+]
+low_severity = mean_aggregated_data_pca[
+    mean_aggregated_data_pca["Metadata_genotype"] == "Mid-Severity"
+]
+wt = mean_aggregated_data_pca[
+    mean_aggregated_data_pca["Metadata_genotype"] == "Wild Type"
+]
+print(len(high_severity), len(low_severity), len(wt))
+
+
+# In[4]:
+
+
+# drop the the Metadata columns
+high_severity = high_severity.drop(columns=["Metadata_genotype", "Metadata_replicate"])
+low_severity = low_severity.drop(columns=["Metadata_genotype", "Metadata_replicate"])
+wt = wt.drop(columns=["Metadata_genotype", "Metadata_replicate"])
+# convert the df to coordinates
+high_severity_coords = high_severity.to_numpy()
+low_severity_coords = low_severity.to_numpy()
+wt_coords = wt.to_numpy()
+print(high_severity_coords.shape, low_severity_coords.shape, wt_coords.shape)
+
+
+# #### High-Severity
+
+# In[5]:
+
+
+# define the mean and the inverse covariance matrix needed for the mahalanobis distance calculation
+cov_matrix = np.cov(high_severity_coords, rowvar=False)
+inv_cov_matrix = np.linalg.inv(cov_matrix)  # Use pseudo-inverse instead of inverse
```

Comment:
Is there a specific reason to use psuedo-inverse here? I would recommend providing a little more clarity here for this choice if possible.

Reply:
Removed psuedo and replaced with inverse